### PR TITLE
Introduced userspace specs

### PR DESCRIPTION
**Pull request by kurtschelfthout**
Few things to work out but putting it out there to discuss

### CODE REVIEW

**Comment by hrichardlee** [comment]
Looks super!

File: src/meadowdb/connection.py [hunk 1]
```diff
@@ -21,52 +20,67 @@
...
-_default_userspace_name: Optional[str] = None
+_default_userspace_spec: Optional[UserspaceSpec] = None
 
 
-def get_default_userspace() -> str:
+def get_default_userspace_spec() -> UserspaceSpec:
     """
-    Gets the default userspace, which can be set via environment variable or by the
-    set_default_userspace function. Whenever userspace is not set, the default userspace
+    Gets the default userspace spec, which can be set via environment variable MEADOWDB_DEFAULT_USERSPACE
```
Comment (hrichardlee) [inline src/meadowdb/connection.py]: I've been formatting comments with https://plugins.jetbrains.com/plugin/7234-wrap-to-column it's a manual process so I definitely miss some stuff but I figure better than leaving these kinds of long lines in?

File: src/meadowdb/connection.py [hunk 2]
```diff
@@ -21,52 +20,67 @@
...
-    if _default_userspace_name is None:
+    global _default_userspace_spec
+    if _default_userspace_spec is None:
         if _MEADOWDB_DEFAULT_USERSPACE in os.environ:
-            _default_userspace_name = os.environ[_MEADOWDB_DEFAULT_USERSPACE]
+            userspace_spec_str = os.environ[_MEADOWDB_DEFAULT_USERSPACE]
+            userspaces = userspace_spec_str.split(",")
+            if not userspaces:
+                raise ValueError(
+                    f"The environment variable {_MEADOWDB_DEFAULT_USERSPACE} is set, but is not a valid userspace spec. Please set it to a comma-separated list of at least one userspace name, for example 'main,test-space'."
```
Comment (hrichardlee) [inline src/meadowdb/connection.py]: Same thing with string constants, these I've been formatting totally manually

File: src/meadowdb/connection.py [hunk 3]
```diff
@@ -181,7 +197,7 @@ def delete_where_equal(
         self,
         table_name: str,
         delete_where_equal_df: pd.DataFrame,
-        userspace: str = None,
+        userspace_spec: UserspaceSpec = None,
```
Comment (hrichardlee) [inline src/meadowdb/connection.py]: Should be `Optional[UserspaceSpec]`

File: src/meadowdb/connection.py [hunk 4]
```diff
@@ -195,27 +211,28 @@ def delete_where_equal(
         This is equivalent to DELETE FROM table_name WHERE (a = 1 AND b = 2) OR (a = 3
         and b = 4)
         """
-        self.write(table_name, None, userspace, delete_where_equal_df)
+        self.write(table_name, None, userspace_spec, delete_where_equal_df)
 
-    def delete_all(self, table_name: str, userspace: str = None) -> None:
+    def delete_all(self, table_name: str, userspace_spec: UserspaceSpec = None) -> None:
```
Comment (hrichardlee) [inline src/meadowdb/connection.py]: Same here, should be `Optional[UserspaceSpec]`

File: src/meadowdb/readerwriter_shared.py [hunk 5]
```diff
@@ -30,9 +30,11 @@
 from __future__ import annotations
 
 import abc
+import enum
```
Comment (hrichardlee) [inline src/meadowdb/readerwriter_shared.py]: Looks like this is not used?

File: src/meadowdb/table_versions_client_local.py [hunk 6]
```diff
@@ -165,7 +165,10 @@ def _add_or_update_table_names(self, table_name: TableName) -> None:
             ] = table_name
 
     def get_current_table_version(
-        self, userspace: str, table_name: str, max_version_number: Optional[int]
+        self,
```
Comment (hrichardlee) [inline src/meadowdb/table_versions_client_local.py]: Unnecessary formatting change?

**Review by hrichardlee — APPROVED** [review]
A few minor comments, feel free to change or not before merging

File: src/meadowdb/connection.py (hunk 1 above)
Comment (kurtschelfthout) [reply src/meadowdb/connection.py]: Ok, I was actually using VSCode, but have now installed Rewrap extension which seems to do the right thing.

File: src/meadowdb/connection.py (hunk 2 above)
Comment (kurtschelfthout) [reply src/meadowdb/connection.py]: Yeah. I'll reformat it manually as I can't find anything that does it for me. My POV is that if it isn't reasonably automatic and checked, it's not really worth paying a lot of attention to.

(total aside, but max length at 79 with current wide screens feels pretty ridiculous haha)

File: src/meadowdb/connection.py (hunk 3 above)
Comment (kurtschelfthout) [reply src/meadowdb/connection.py]: Was wondering why mypy didn't complain, turns out there is a setting `no_implicit_optional`. I'll enable it and fix.

File: src/meadowdb/readerwriter_shared.py (hunk 5 above)
Comment (kurtschelfthout) [reply src/meadowdb/readerwriter_shared.py]: Indeed, fixed.

File: src/meadowdb/table_versions_client_local.py (hunk 6 above)
Comment (kurtschelfthout) [reply src/meadowdb/table_versions_client_local.py]: Fixed
